src/engine/mod.rs

```diff
@@ -0,0 +1,320 @@
+use self::context::Context;
+use self::equal::RelativeEq;
+use crate::{
+    builtin::{
+        filter::{self, FilterFn},
+        function::{self, FunctionFn},
+    },
+    error::{bail, Result},
+    parser::ast::*,
+    utils::validate_f64,
+};
+use serde_json::{Number, Value};
+use std::collections::HashMap;
+
+pub mod context;
+mod equal;
+
+pub struct Engine {
+    functions: HashMap<String, FunctionFn>,
+    filters: HashMap<String, FilterFn>,
+}
+
+impl Default for Engine {
+    fn default() -> Engine {
+        let mut filters = HashMap::new();
+        filters.insert("upper".into(), Box::new(filter::upper) as FilterFn);
+        filters.insert("lower".into(), Box::new(filter::lower) as FilterFn);
+
+        let mut functions = HashMap::new();
+        functions.insert("uuidv4".into(), Box::new(function::uuidv4) as FunctionFn);
+
+        Engine { functions, filters }
+    }
+}
+
+impl Engine {
+    fn eval_math(&self, lhs: &Number, rhs: &Number, operator: MathOperator) -> Result<Number> {
+        match operator {
+            MathOperator::Addition => {
+                if lhs.is_i64() && rhs.is_i64() {
+                    if let Some(x) = lhs.as_i64().unwrap().checked_add(rhs.as_i64().unwrap()) {
```

Comment:
Right now the `if let` skips the `None` variant checking which signifies the integer overflow on addition.
While unlikely - may be worth forwarding the error of the overflow condition to the user here, or even panic'ing. - what do you think ?

Reply:
Disagree. I do not want to panic at all. I consider panicking as a bad behavior and no library should panic, especially when we're talking about user input here.

The decission I made here is that I try to do float math if i64 fails and then fail. Don't have strong opinion about this.